### DISCUSSION TITLE
Feature request : configurable shortcuts for triggering scripts on host

### DISCUSSION THREAD

**Discussion by ghost** [input-leap/input-leap#1784] [Feature Requests]
> This issue has been migrated from old Barrier Github repository https://github.com/debauchee/barrier/issues/1079
> 
> Issue created on: 2021-02-26 by @Potomac
> Issue last updated on: 2021-02-26

Hello,

a PC monitor can have several inputs : vga, hdmi, dvi, display port,

I decided to connect 2 PCs on my iiyama monitor (VGA and DVI ports),
I want to select the input (VGA, DVI) by sending DDC commands to the monitor, instead of using the OSD of my monitor, it is possible by using DDC protocol : 
https://en.wikipedia.org/wiki/Display_Data_Channel

I can write a bash script with inside commands made by "ddcutil", that can automatically switch the monitor to the desired video input (VGA, DVI) : 
https://passthroughpo.st/switching-monitor-inputs-software-ddc/
https://www.ddcutil.com/

The feature I want for barrier : have the ability to trigger a custom bash script with the keyboard, by configuring shortcuts, the bash script would run the appropriate ddcutil command for switching the monitor to the correct input (VGA, DVI etc...), according to the PC I want to control.

Thanks.

**Comment by eldorel** [comment]
(Note: I replied to this on the barrier issue first, but copied it here since this seems to be the more active fork now)

@Potomac I have no idea where I found this, but barrier has a command line flag for the server that can be used to trigger a script on screen change. 

`--screen-change-script /path/to/script` 

This runs the named script on switch with the name of the screen as the first argument. 
I have a bash script with a case statement that changes the color of my RGB keyboard to reflect which system i'm currently controlling. 

--------

That said, I do actually have a different use case for an 'action' that can be triggered by keypresses to execute a script on the host (and hopefully allow this as a secondary action for that keypress).  

I want to configure a single shortcut on all of my systems to toggle screen brightness (also using DDCutil actually, and a different tool for the windows clients  ) and have barrier send that keypress to everything. 

Sending the key works, but the shortcut system in X11 runs at a lower level than barrier and hides the keypress, so I actually _can't_ have barrier trigger a keypress on an action that's linked to an X11 shortcut on the host.   

Either the shortcut doesn't see the key injected by barrier, or the shortcut hijacks the key and prevent barrier from registering the key with x11.  (This is a shortcoming with X, only one tool can register a global shortcut for a given key)

I also tried changing the sent key to be different for the shortcut, but again the registered shortcut doesn't see keys sent by barrier. 
example: `keystroke(F21) = keystroke(F22,*)`


Being able to setup something similar to this: 
`keystroke(F21) = keystroke(F21,*)+run(/path/to/script)` 
or
`keystroke(F21) = run(/$home/relative/path/to/script,*)` 
would allow me to execute the brightness control script on the server or trigger a script on all clients in the same location.

**Comment by xpufx** [comment]
@eldorel how to you initiate the server/client with that command line? I am using the gui and don't see a way to easily add it. That switch didn't come up in the github code search either. Are you sure it exists?

**Comment by eldorel** [comment]
> @eldorel how to you initiate the server/client with that command line? 
With a modified shortcut in the OS program launcher on linux. 
I am not using the automatically added service launch.

> I am using the gui and don't see a way to easily add it. That switch didn't come up in the github code search either. Are you sure it exists?

Original pull:
https://github.com/debauchee/barrier/pull/249
Imported here:
https://github.com/input-leap/input-leap/pull/249

I use it hundreds of times a day along with a custom script to trigger openRGB to change my RGB lighting depending on the system I am controlling.

**Comment by xpufx** [comment]
Thanks I figured out part of it. I was using the GUI. I copied the barriers command line from it and added the switch paramater. I was able to test switching to my windows screen. Now I have to figure out how to come back from it. Thanks.

**Comment by xpufx** [comment]
I would never figure out the hostname part without that script example and it turns out I don't have to figure out the windows part because that's all happening on the barriers side. This is working beautifully. Thanks a lot.
